File: src/main/kotlin/com/xs0/gqlktx/types/gql/GObjectType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.QueryException;
import com.xs0.gqlktx.schema.builder.TypeKind;
import com.xs0.gqlktx.schema.intro.GqlIntroType;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.*;

public class GObjectType extends GFieldedType {
    private Set<GInterfaceType> interfaces = Collections.emptySet(); // lazy allocate

    public GObjectType(String name, Map<String, GField> fields) {
        super(name, fields);
    }

    @Override
    public TypeKind getKind() {
        return TypeKind.OBJECT;
    }

    @Override
    public boolean validAsArgumentType() {
        return false;
    }

    @Override
    public void coerceValue(JsonObject raw, String key, JsonObject out) throws QueryException {
        throw new QueryException("Union type " + getName() + " can't be used as a variable");
    }

    @Override
    public void coerceValue(JsonArray raw, int index, JsonArray out) throws QueryException {
        throw new QueryException("Union type " + getName() + " can't be used as a variable");
    }

    @Override
    public void toString(StringBuilder sb) {
        sb.append("type ").append(getName());

        boolean first = true;
        for (GInterfaceType i : interfaces) {
            sb.append(first ? " implements " : ", ");
            first = false;
            sb.append(i.getName());
        }

        sb.append(" {\n");
        dumpFieldsToString(sb);
        sb.append("}\n");
    }

    void addInterface(GInterfaceType interfaceType) {
        if (interfaces.isEmpty())
            interfaces = new LinkedHashSet<>();

        interfaces.add(interfaceType);
    }

    public List<GqlIntroType> getInterfacesForIntrospection() {
        ArrayList<GqlIntroType> res = new ArrayList<>(interfaces.size());
        for (GInterfaceType inter : interfaces)
            res.add(inter.introspector());
        return res;
    }
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GObjectType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.QueryException
import com.xs0.gqlktx.dom.Value
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.schema.intro.GqlIntroType

import java.util.*

class GObjectType(name: String, fields: Map<String, GField>, description: String?) : GFieldedType(name, fields, description) {

    override val kind: TypeKind
        get() = TypeKind.OBJECT

    override fun coerceValue(raw: Value): Value {
        throw QueryException("Object type $name can't be used as a variable")
    }

    override fun toString(sb: StringBuilder) {
        sb.append("type ").append(name)

        var first = true
        for (i in getInterfaces()) {
            sb.append(if (first) " implements " else ", ")
            first = false
            sb.append(i.name)
        }

        sb.append(" {\n")
        dumpFieldsToString(sb)
        sb.append("}\n")
    }

    val interfacesForIntrospection: List<GqlIntroType> by lazy {
        val res = ArrayList<GqlIntroType>(getInterfaces().size)
        for (inter in getInterfaces())
            res.add(inter.introspector)
        res
    }
}
